File: dotNET/Q/Trading/Bar.cs
using System;
using System.Collections.Generic;
using java.lang;
using Q.Util;
using Double=System.Double;
using JBar = systemdb.data.Bar;
using JTick = systemdb.data.Tick;
using Math=System.Math;

namespace Q.Trading {
    public class Bar : Objects, IEquatable<Bar> {

        public readonly double open;
        public readonly double high;
        public readonly double low;
        public readonly double close;
        public readonly double volume;
        public readonly double openInterest;
        public readonly DateTime time;

        public Bar(JBar bar) : this(bar, true) {}

        public Bar(JBar bar, bool strict) : this(bar.open(), bar.high(), bar.low(), bar.close(), date(bar.date()), strict) {
            var jOpenInterest = bar.openInterest();
            openInterest = jOpenInterest == null ? Double.NaN : jOpenInterest.doubleValue();
            var jVolume = bar.volume();
            volume = jVolume == null ? Double.NaN : jVolume.doubleValue();
        }

        public Bar(double open, double high, double low, double close) : this(open, high, low, close, true) {}
        public Bar(double open, double high, double low, double close, bool strict) : this(open, high, low, close, default(DateTime), strict) {}

        public Bar(double open, double high, double low, double close, DateTime time) : this(open, high, low, close, time, true) {}
        public Bar(double open, double high, double low, double close, DateTime time, bool strict) {
            this.open = open;
            this.high = high;
            this.low = low;
            this.close = close;
            this.time = time;
            if (strict) checkBar();
        }

        public Bar(systemdb.data.Tick tick) : this(tick.open, tick.high, tick.low, tick.last, date(tick.time)) {}

        Bar(double open, double high, double low, double close, DateTime time, double volume, double openInterest) 
            : this(open, high, low, close, time) {
            this.volume = volume;
            this.openInterest = openInterest;
        }

        public Bar(double testData) : this(testData, testData, testData, testData) {}
        public Bar(double testData, DateTime time) : this(testData, testData, testData, testData, time) {}

        public void checkBar() {
            Bomb.unless(high >= low, () => "high " + paren(high) + " must be above low " + paren(low) + "\nbar: " + this);
        }

        public double range() {
            return high - low;
        }

        public Bar update(Tick tick) {
            return new Bar(open, Math.Max(tick.price, high), Math.Min(tick.price, low), tick.price, tick.time, volume, openInterest);
        }

        public Bar update(DateTime newTime) {
            return new Bar(open, high, low, close, newTime, volume, openInterest);
        }

        public override string ToString() {
            return paren(commaSep(ymdHuman(time), open, high, low, close, volume, openInterest));
        }

        public bool Equals(Bar bar) {
            if (bar == null) return false;
            if (open != bar.open) return false;
            if (high != bar.high) return false;
            if (low != bar.low) return false;
            if (close != bar.close) return false;
            if (volume != bar.volume) return false;
            return openInterest == bar.openInterest && Equals(time, bar.time);
        }

        public override bool Equals(object obj) {
            return ReferenceEquals(this, obj) || Equals(obj as Bar);
        }

        public override int GetHashCode() {
            var result = open.GetHashCode();
            result = 29 * result + high.GetHashCode();
            result = 29 * result + low.GetHashCode();
            result = 29 * result + close.GetHashCode();
            result = 29 * result + volume.GetHashCode();
            result = 29 * result + openInterest.GetHashCode();
            result = 29 * result + time.GetHashCode();
            return result;
        }

        public List<double> orderedHighLow() {
//          This is a rounded version of: open - low <= high - open
            return 2 * open - low - high < 1e-10
                    ? list(low, high)
                    : list(high, low);
        }

        public JTick jTick() {
            return new JTick(close, (long) volume, open, high, low, jDate(time));
        }

        public Tick tick() {
            return new Tick(close, (ulong) volume, time);
        }

        public JBar java() {
            return new JBar(jDate(time), open, high, low, close, new Long((long) volume), new Long((long) openInterest));
        }
    }
}

File: Java/systematic/src/systemdb/data/Bar.java
package systemdb.data;

import static util.Strings.*;

import java.io.*;
import java.util.*;

import util.*;

public class Bar implements Comparable<Bar>, Serializable {

    private static final long serialVersionUID = 1L;
    private static final long NA = Long.MIN_VALUE + 123;
    private final Date date;
    private final double open;
    private final double high;
    private final double low;
    private final double close;
    private final Long volume;
    private final Long openInterest;

    public Bar(Date date, double open, double high, double low, double close, Long volume, Long openInterest) {
        this.date = date;
        this.open = open;
        this.high = high;
        this.low = low;
        this.close = close;
        this.volume = volume;
        this.openInterest = openInterest;
    }

    @SuppressWarnings("unchecked") @Override public String toString() {
        return paren(commaSep(strings(date, open, high, low, close, volume, openInterest)));
    }

    @Override public int hashCode() {
        final int prime = 31;
        int result = 1;
        long temp;
        temp = Double.doubleToLongBits(close);
        result = prime * result + (int) (temp ^ (temp >>> 32));
        result = prime * result + ((date == null) ? 0 : date.hashCode());
        temp = Double.doubleToLongBits(high);
        result = prime * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(low);
        result = prime * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(open);
        result = prime * result + (int) (temp ^ (temp >>> 32));
        result = prime * result + ((openInterest == null) ? 0 : openInterest.hashCode());
        result = prime * result + ((volume == null) ? 0 : volume.hashCode());
        return result;
    }



    @Override public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass()) return false;
        final Bar other = (Bar) obj;
        if (Double.doubleToLongBits(close) != Double.doubleToLongBits(other.close)) return false;
        if (date == null) {
            if (other.date != null) return false;
        } else if (!date.equals(other.date)) return false;
        if (Double.doubleToLongBits(high) != Double.doubleToLongBits(other.high)) return false;
        if (Double.doubleToLongBits(low) != Double.doubleToLongBits(other.low)) return false;
        if (Double.doubleToLongBits(open) != Double.doubleToLongBits(other.open)) return false;
        if (openInterest == null) {
            if (other.openInterest != null) return false;
        } else if (!openInterest.equals(other.openInterest)) return false;
        if (volume == null) {
            if (other.volume != null) return false;
        } else if (!volume.equals(other.volume)) return false;
        return true;
    }



    public boolean in(Range range) {
        return range.containsInclusive(date);
    }

    public Date date() {
        return date;
    }
    
    public double open() {
        return open;
    }

    public double close() {
        return close;
    }
    
    public double high() {
        return high;
    }
    
    public double low() {
        return low;
    }
    
    public Long openInterest() {
        return openInterest;
    }
    
    public Long volume() {
        return volume;
    }

    public long rOpenInterest() {
        return openInterest == null ? NA : openInterest;
    }
    
    public long rVolume() {
        return openInterest == null ? NA : volume;
    }
    
    
    public static long na() {
        return NA;
    }

    @Override public int compareTo(Bar o) {
        return date.compareTo(o.date);
    }
}
